keps/sig-storage/3314-csi-changed-block-tracking/README.md

```diff
@@ -0,0 +1,1692 @@
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### The CSI SnapshotMetadata Service API
+
+> In this section we use the terminology of the gRPC specification where the
+> word *service* is assumed to be a *gRPC service* and not a Kubernetes service,
+> while the word *plugin* is
+> used to refer to the software component that implements the gRPC service.
+
+The CSI specification will be extended with the addition of the following new, optional
+**SnapshotMetadata** [gRPC service](https://grpc.io/docs/what-is-grpc/core-concepts/#service-definition).
+The [SP Snapshot Metadata](#the-sp-snapshot-metadata-service) plugin implements
+this service.
+
+The service is defined as follows, and will be described in the sub-sections below.
+Refer to [CSI PR 551](https://github.com/container-storage-interface/spec/pull/551)
+for the official specification. 
+```
+service SnapshotMetadata {
+  rpc GetMetadataAllocated(GetMetadataAllocatedRequest)
+    returns (stream GetMetadataAllocatedResponse) {}
+  rpc GetMetadataDelta(GetMetadataDeltaRequest)
+    returns (stream GetMetadataDeltaResponse) {}
+}
+
+enum BlockMetadataType {
+  UNKNOWN=0;
+  FIXED_LENGTH=1;
+  VARIABLE_LENGTH=2;
+}
+
+message BlockMetadata {
+  int64 byte_offset = 1;
+  int64 size_bytes = 2;
+}
+
+message GetMetadataAllocatedRequest {
+  string snapshot_id = 1;
+  int64 starting_offset = 2;
+  int32 max_results = 3;
+  map<string, string> secrets = 4;
+}
+
+
+message GetMetadataAllocatedResponse {
+  BlockMetadataType block_metadata_type = 1;
+  int64 volume_capacity_bytes = 2;
+  repeated BlockMetadata block_metadata = 3;
+}
+
+message GetMetadataDeltaRequest {
+  string base_snapshot_id = 1;
+  string target_snapshot_id = 2;
+  int64 starting_offset = 3;
+  int32 max_results = 4;
+  map<string, string> secrets = 5;
+}
+
+message GetMetadataDeltaResponse {
+  BlockMetadataType block_metadata_type = 1;
+  int64 volume_capacity_bytes = 2;
+  repeated BlockMetadata block_metadata = 3;
+}
+```
+
+#### Metadata Format
+Block volume data ranges are specified by a sequence of `(ByteOffset, Length)` tuples,
+with the tuples in ascending order of `ByteOffset` and no overlap between adjacent tuples.
+There are two prevalent styles, *extent-based* or *block-based*,
+which describe if the `Length` field of the tuples in a sequence can
+**vary** or are **fixed** across all the tuples in the sequence.
+The **SnapshotMetadata** service permits either style at the discretion of the plugin,
+and it is required that a client of this service be able to handle both styles.
+
+The `BlockMetadataType` enumeration specifies the style used: `FIXED_LENGTH` or `VARIABLE_LENGTH`.
+When the *block-based* style (`FIXED_LENGTH`) is used it is up to the SP plugin to define the
+block size.
```

Comment:
How will backup clients make use of `FIXED_LENGTH`/`VARIABLE_LENGTH`?

Reply:
A backup client would, in general, have to map the "view" of the block device returned by the API to its own view of the block device. For example, a client could have its own concept of block size and would have to map, say, a VMware extent map, to a range of blocks; conversely, it may have to map a range of AWS EBS blocks to an extent map.

Reply:
The purpose of `FIXED_LENGTH`/`VARIABLE_LENGTH` is still not clear to me.

The API returns an ascending, non-overlapping sequence of `BlockMetadata` elements. Backup clients may accumulate adjacent `BlockMetadata` elements (i.e. convert from blocks to extents) and then align offsets/lengths to the backup client's internal block size, but this does not require `BlockMetadataType`.

Can you give a concrete example of how a backup client will use `BlockMetadataType` to clarify?

Reply:
When we discussed this in the CSI WG, the argument was thus:
* If the SP uses FIXED_LENGTH, it's a promise that lengths will be constant for the given pair of snapshots, which may unlock certain optimizations on the backup software side which wouldn't be possible with VARIABLE_LENGTH
* If the SP uses VARIABLE_LENGTH, it remains free to return whatever lengths it wants -- possible folding consecutive change blocks and thereby compressing the metadata, and the backup software has to cope with it.

If these details aren't clear from the CSI spec, then it's a failure of spec writing, and we should amend the spec to be more clear about the exact semantics of that field.

Reply:
@bswartz While the spec covers the effect of `BlockMetadataType` on the `Length` field, it's still unclear to me how a backup client would use this information.

The one example I can think of, making decisions about backup data layout (e.g. block size and the backup client's internal metadata parameters), doesn't seem to be well-served by `BlockMetadataType`:

`BlockMetadataType` only applies for the current API call. Maybe it could even change between successive calls on the same snapshot pair. Definitely between different snapshot pairs from the same volume.

Is a backup client that uses this information making the assumption that `BlockMetadataType` and the block size does not change? If yes, then either backup clients cannot use `BlockMetadataType` for this purpose or the API design needs adjustment.

An API that captures permanent attributes like block size would be something like `GetVolumeMetadataHints(volume_id) -> VolumeMetadataHints` where the spec guarantees the stability of the hints across snapshots of the volume.

Or maybe I'm still missing how backup clients could use `BlockMetadataType`?

Reply:
Recall that these new snapshot metadata APIs are stream-oriented, so "one API call" is actually the whole stream of metadata for a given pair of snapshots. If the SP commits to always use a Length of 4096 (for example) then the backup software can safely allocate a fixed-size bitmap where each bit represents 4k of change for the whole snasphot.

If an SP returns variable length blocks, then the backup software has to be prepared to use something other than a bitmap to represent the delta, or be prepared to guess a optimal bitmap granularity and be prepared to adjust delta blocks to fit the bitmap, possibly wasting small amounts of space by backing up unchanged data.

Reply:
`FIXED_LENGTH` is expensive over the wire because adjacent `BlockMetadata` elements cannot be sent merged. It would be more efficient for the SP to include an optional block size hint field in the response and allow the `Length` field to be vary instead of having `BlockMetadataType`.

Reply:
Good thing the API is alpha and we can evolve it. I agree this isn't the most efficient transport mechanism, but it's dramatically more efficient that what's being done in the absence of this feature. The designers were optimizing for simplicity rather than efficiency, which is the right place to focus in an alpha version of anything.

Reply:
Thanks for considering this feedback!